### PR TITLE
BUG: Add np.uintc to _factorizers in merge.py to fix KeyError when merging DataFrames with uintc columns

### LINKED ISSUE #58713: BUG: pd.merge fail with numpy.uintc on Windows
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [ ] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.

### Reproducible Example

This bug is fixed by adding np.uintc to _factorizers in pd.merge. Could someone upload it fixed please?.  For me the bug is fixed by adding in line 114 of core/reshape/merge.py file: np.uintc:libhastable.UInt32Factorizer. 

```python
import numpy as np
import pandas as pd

df1 = pd.DataFrame({'a':['foo','bar'],'b':np.array([1,2], dtype=np.uintc)})
df2 = pd.DataFrame({'a':['foo','baz'],'b':np.array([3,4], dtype=np.uintc)})
df3=df1.merge(df2, how = 'outer')
print(df3)
```

### Issue Description

Traceback (most recent call last):
  File "C:\Users\noname37486\PycharmProjects\pythonProject3\main.py", line 10, in <module>
    df3=df1.merge(df2, how = 'outer')
  File "C:\Users\noname37486\PycharmProjects\pythonProject3\venv\lib\site-packages\pandas\core\frame.py", line 10832, in merge
    return merge(
  File "C:\Users\noname37486\PycharmProjects\pythonProject3\venv\lib\site-packages\pandas\core\reshape\merge.py", line 184, in merge
    return op.get_result(copy=copy)
  File "C:\Users\noname37486\PycharmProjects\pythonProject3\venv\lib\site-packages\pandas\core\reshape\merge.py", line 886, in get_result
    join_index, left_indexer, right_indexer = self._get_join_info()
  File "C:\Users\noname37486\PycharmProjects\pythonProject3\venv\lib\site-packages\pandas\core\reshape\merge.py", line 1151, in _get_join_info
    (left_indexer, right_indexer) = self._get_join_indexers()
  File "C:\Users\noname37486\PycharmProjects\pythonProject3\venv\lib\site-packages\pandas\core\reshape\merge.py", line 1125, in _get_join_indexers
    return get_join_indexers(
  File "C:\Users\noname37486\PycharmProjects\pythonProject3\venv\lib\site-packages\pandas\core\reshape\merge.py", line 1740, in get_join_indexers
    zipped = zip(*mapped)
  File "C:\Users\v\PycharmProjects\pythonProject3\venv\lib\site-packages\pandas\core\reshape\merge.py", line 1737, in <genexpr>
    _factorize_keys(left_keys[n], right_keys[n], sort=sort)
  File "C:\Users\noname37486\PycharmProjects\pythonProject3\venv\lib\site-packages\pandas\core\reshape\merge.py", line 2539, in _factorize_keys
    klass, lk, rk = _convert_arrays_and_get_rizer_klass(lk, rk)
  File "C:\Users\noname37486\PycharmProjects\pythonProject3\venv\lib\site-packages\pandas\core\reshape\merge.py", line 2616, in _convert_arrays_and_get_rizer_klass
    klass = _factorizers[lk.dtype.type]
KeyError: <class 'numpy.uintc'>

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit                : d9cdd2ee5a58015ef6f4d15c7226110c9aab8140
python                : 3.9.6.final.0
python-bits           : 64
OS                    : Windows
OS-release            : 10
Version               : 10.0.22631
machine               : AMD64
processor             : Intel64 Family 6 Model 165 Stepping 2, GenuineIntel
byteorder             : little
LC_ALL                : None
LANG                  : None
LOCALE                : es_ES.cp1252

pandas                : 2.2.2
numpy                 : 1.26.4
pytz                  : 2024.1
dateutil              : 2.9.0.post0
setuptools            : 57.0.0
pip                   : 21.1.2
Cython                : None
pytest                : None
hypothesis            : None
sphinx                : None
blosc                 : None
feather               : None
xlsxwriter            : None
lxml.etree            : None
html5lib              : None
pymysql               : None
psycopg2              : None
jinja2                : None
IPython               : None
pandas_datareader     : None
adbc-driver-postgresql: None
adbc-driver-sqlite    : None
bs4                   : None
bottleneck            : None
dataframe-api-compat  : None
fastparquet           : None
fsspec                : None
gcsfs                 : None
matplotlib            : None
numba                 : None
numexpr               : None
odfpy                 : None
openpyxl              : None
pandas_gbq            : None
pyarrow               : None
pyreadstat            : None
python-calamine       : None
pyxlsb                : None
s3fs                  : None
scipy                 : None
sqlalchemy            : None
tables                : None
tabulate              : None
xarray                : None
xlrd                  : None
zstandard             : None
tzdata                : 2024.1
qtpy                  : None
pyqt5                 : None

</details>

### PR DESCRIPTION
**Pull request by Tirthchoksi22**
- [x] closes #58713 (Replace xxxx with the GitHub issue number)
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

**Comment by Tirthchoksi22** [comment]
hiii @myles the PR is ready to merge

**Comment by Tirthchoksi22** [comment]
Yes, it appears that there is indeed a regression on Windows machines. The issue seems to have resurfaced after being resolved in a previous release.

> thanks @Tirthchoksi22
> 
> the same issue was fixed for `np.intc` in #53175.
> 
> This is only for Windows and is it a regression from a previous release?
> 
> if so could probably use the previous fix/tests/release note for this PR?

**Comment by Tirthchoksi22** [comment]
@simonjayhawkins After reviewing the code and considering the previous fix, it appears that the solution implemented in this pull request is indeed identical to the one used in the previous resolution of the regression.

> thanks @Tirthchoksi22
> 
> the same issue was fixed for `np.intc` in #53175.
> 
> This is only for Windows and is it a regression from a previous release?
> 
> if so could probably use the previous fix/tests/release note for this PR?

**Comment by Tirthchoksi22** [comment]
@simonjayhawkins Also guide me what to do next do i have to create new PR with previous fix/tests/release or this is ok ??as this would be my first Open Source Contribution

**Comment by simonjayhawkins** [comment]
you can make changes to you branch and push those changes to update this PR. No need to close and open a new PR.

All bug fixes and regression fixes need a test to ensure that the issue does not resurface here. In this case, looking at the fix for `np.intc` the parameterization of `test_join_multi_dtypes` was updated. Perhaps could do the same here?

For the release note, if you (I can't check a windows only bug easily) determine for which pandas release it was a regression, or if that is many releases ago, then we maybe need a note either under the bug fixes or regression section of the release notes. If a bug fix would go in the 3.0 what's new. If a regression, would go in the 2.2.x release notes (but that depends if we are doing another patch release)

cc @lithomas1

**Comment by simonjayhawkins** [comment]
ci is failing, 

/pre-commit

**Comment by Tirthchoksi22** [comment]
> ci is failing,
> 
> /pre-commit


so what to do now ??

**Comment by simonjayhawkins** [comment]
I thought that comment should have triggered the bot.

**Comment by simonjayhawkins** [comment]
@github-actions pre-commit

**Comment by Tirthchoksi22** [comment]
> /pre-commit

I dont think there's a command like this

**Comment by simonjayhawkins** [comment]
pre-commit.ci autofix

File: pandas/tests/reshape/merge/test_merge.py
Comment (simonjayhawkins) [inline pandas/tests/reshape/merge/test_merge.py:1498]: So it appears that this has been updated since the last fix.

I doubt that this should be changed. it's probably that `any_int_numpy_dtype` (and therefore `tm.UNSIGNED_INT_NUMPY_DTYPES` ) should be updated.

However, `tm.UNSIGNED_INT_NUMPY_DTYPES` does not appear to contain `np.intc` either. (maybe how the regression was uncaught) 

So looks like will need to add both `np.intc` and `np.uintc` to `tm.UNSIGNED_INT_NUMPY_DTYPES`

Comment (Tirthchoksi22) [reply pandas/tests/reshape/merge/test_merge.py:1498]: ok i will add it in `tm.UNSIGNED_INT_NUMPY_DTYPES`

Comment (simonjayhawkins) [reply pandas/tests/reshape/merge/test_merge.py:1498]: Can you explain the issue to me (as I don't have the platform to test)

Although my previous comment was to align better with the previous fix, I don't know why we need to explicitly cater for `np.intc` and `np.uintc` since we don't need to for the other Python API “C-like” names such as `numpy.short`.

So adding these does not really make sense to me as I don't really understand the issue.

Comment (Tirthchoksi22) [reply pandas/tests/reshape/merge/test_merge.py:1498]: @simonjayhawkins what i think is the issue arises because certain functionalities within `pandas` were not properly handling `np.intc` and `np.uintc` data types. As a result, when these data types were encountered, it led to unexpected keyerrors.that's what i feel but don't know why we don't need to mention this in other Python API's

Comment (Tirthchoksi22) [reply pandas/tests/reshape/merge/test_merge.py:1498]: I still didnt understand why regression occur on windows side

**Comment by Tirthchoksi22** [comment]
@simonjayhawkins Could you kindly confirm if the changes proposed in this pull request are satisfactory, or if there are any additional adjustments needed before merging?

File: pandas/tests/reshape/merge/test_merge.py
Comment (simonjayhawkins) [inline pandas/tests/reshape/merge/test_merge.py:1496]: this parameterization (d1) does not have `np.uintc`!

**Review by simonjayhawkins — CHANGES_REQUESTED** [review]
needs a release note. put in 3.0.0 notes for now.

**Comment by Tirthchoksi22** [comment]
pre-commit.ci autofix

**Comment by cmjcharlton** [comment]
I was browsing recent pull requests and have a Windows machine so thought I would give this a quick test for you. It looks like the difference is that on Windows np.uintc is not aliased to any of the types already in the _factorizers dictionary:

```python
Python 3.12.3 (tags/v3.12.3:f6650f9, Apr  9 2024, 14:05:25) [MSC v.1938 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import numpy as np
>>> np.uintc is np.int64
False
>>> np.uintc is np.longlong
False
>>> np.uintc is np.int32
False
>>> np.uintc is np.int16
False
>>> np.uintc is np.int8
False
>>> np.uintc is np.uint64
False
>>> np.uintc is np.uint32
False
>>> np.uintc is np.uint16
False
>>> np.uintc is np.uint8
False
>>> np.uintc is np.bool_
False
>>> np.uintc is np.float64
False
>>> np.uintc is np.float32
False
>>> np.uintc is np.complex64
False
>>> np.uintc is np.complex128
False
>>> np.uintc is np.object_
False
```

Whereas at least in Linux it is aliased to the np.uint32 type:
```python
Python 3.12.3 (main, Apr 10 2024, 05:33:47) [GCC 13.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import numpy as np
>>> np.uintc is np.int64
False
>>> np.uintc is np.longlong
False
>>> np.uintc is np.int32
False
>>> np.uintc is np.int16
False
>>> np.uintc is np.int8
False
>>> np.uintc is np.uint64
False
>>> np.uintc is np.uint32
True
>>> np.uintc is np.uint16
False
>>> np.uintc is np.uint8
False
>>> np.uintc is np.bool_
False
>>> np.uintc is np.float64
False
>>> np.uintc is np.float32
False
>>> np.uintc is np.complex64
False
>>> np.uintc is np.complex128
False
>>> np.uintc is np.object_
False
```

This is probably related to Windows using the LLP64 64-bit data model vs LP64 used elsewhere (https://en.wikipedia.org/wiki/64-bit_computing#64-bit_data_models). https://stackoverflow.com/questions/76155091/np-uint32-np-uintc-on-windows gives a possible explanation.

**Comment by Tirthchoksi22** [comment]
@simonjayhawkins why this 1 test failed

**Comment by Tirthchoksi22** [comment]
pre-commit.ci autofix

**Comment by Tirthchoksi22** [comment]
@simonjayhawkins this one error is not going help me to get through it

**Comment by Tirthchoksi22** [comment]
@simonjayhawkins your requested changes are done and this PR is ready to merge now

File: doc/source/whatsnew/v3.0.0.rst
Comment (mroeschke) [inline doc/source/whatsnew/v3.0.0.rst:124]: Can you make this a 1 line changelong entry in the `Bug Fixes` section under `Reshaping`?

File: pandas/tests/reshape/merge/test_merge.py
Comment (mroeschke) [inline pandas/tests/reshape/merge/test_merge.py:1495]: Could you keep this test the same and add a new separate test for `np.uintc`?

File: doc/source/whatsnew/v3.0.0.rst
Comment (Tirthchoksi22) [reply doc/source/whatsnew/v3.0.0.rst:124]: Screenshot 2024-05-16 220504
so i have to add 1 line changelong entry in this section but here there's nothing like `Bug Fixes` section

File: pandas/core/reshape/merge.py
Comment (WillAyd) [inline pandas/core/reshape/merge.py:134]: What is uintc a typedef for? This seems strange that we check if it is not uint32 but then still use the UInt32Factorizer anyway

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v3.0.0.rst
Comment (Tirthchoksi22) [reply doc/source/whatsnew/v3.0.0.rst:124]: So you are asking me to make a section called `Bug Fixes` under `Reshaping` section and add a 1 line changelong entry instead of this right ??

File: pandas/tests/reshape/merge/test_merge.py
Comment (Tirthchoksi22) [reply pandas/tests/reshape/merge/test_merge.py:1495]: ok sure

File: pandas/core/reshape/merge.py
Comment (Tirthchoksi22) [reply pandas/core/reshape/merge.py:134]: there should be UInt64Factorizer right ???

File: doc/source/whatsnew/v3.0.0.rst
Comment (mroeschke) [reply doc/source/whatsnew/v3.0.0.rst:124]: i.e. replace this line with a 1 line changelog entry (similarly formatted like the others around it) https://github.com/pandas-dev/pandas/blob/c121940765260b3208f012dad5ccd68a83e3f081/doc/source/whatsnew/v3.0.0.rst?plain=1#L476

File: pandas/core/reshape/merge.py
Comment (WillAyd) [reply pandas/core/reshape/merge.py:134]: Probably. Its hard to say for sure though without knowing what np.uintc is supposed to be

**Comment by Tirthchoksi22** [comment]
pre-commit.ci autofix

**Comment by Tirthchoksi22** [comment]
pre-commit.ci autofix

**Comment by Tirthchoksi22** [comment]
pre-commit.ci autofix

**Comment by simonjayhawkins** [comment]
@Tirthchoksi22 on a procedural note, the `pre-commit.ci autofix` is useful for maintainers to correct the code. As the PR author, you can install pre-commit in your local dev environment to check the code changes you make before pushing them to the PR.

**Comment by Tirthchoksi22** [comment]
> @Tirthchoksi22 on a procedural note, the `pre-commit.ci autofix` is useful for maintainers to correct the code. As the PR author, you can install pre-commit in your local dev environment to check the code changes you make before pushing them to the PR.

ok sorry I didn't know about that

**Comment by simonjayhawkins** [comment]
> Can you please use that comment now because i didnt have install pre-commit in my local environment right now

feel free to issue that command yourself. I was not suggesting that you could not use it. (However, if you look at the pandas documentation, you will see a section for setting up a development environment)

File: pandas/core/reshape/merge.py
Comment (WillAyd) [inline pandas/core/reshape/merge.py:133]:
```suggestion
if np.uintc(0).itemsize == 8:
```

I am wary of trying to solve this through `is` checks - it would be better to more explicitly check the data type size

**Review by WillAyd — CHANGES_REQUESTED** [review]

**Comment by Tirthchoksi22** [comment]
image
here the error is is that d2 is not defined but i have specifically defiend it can you please help me out 
image

**Comment by Tirthchoksi22** [comment]
pre-commit.ci autofix

File: doc/source/whatsnew/v3.0.0.rst
Comment (simonjayhawkins) [inline doc/source/whatsnew/v3.0.0.rst:477]: can you format like the notes around it. 

start with "Bug in ...", keep the description short (one line) but be specific (mention "Windows"), and end with ```(:issue:`58713`)```

Comment (Tirthchoksi22) [reply doc/source/whatsnew/v3.0.0.rst:477]: ok done

Comment (Tirthchoksi22) [reply doc/source/whatsnew/v3.0.0.rst]: ok done

File: pandas/core/reshape/merge.py
Comment (cmjcharlton) [reply pandas/core/reshape/merge.py]: On my Windows machine np.uintc has an itemsize of 4:
```python
Python 3.12.3 (tags/v3.12.3:f6650f9, Apr  9 2024, 14:05:25) [MSC v.1938 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import numpy as np
>>> np.uintc(0).itemsize
4
```
So wouldn't this prevent the proposed fix from being applied?

**Comment by Tirthchoksi22** [comment]
pre-commit.ci autofix

**Comment by Tirthchoksi22** [comment]
@simonjayhawkins @WillAyd @mroeschke Guys any update ???

**Comment by Tirthchoksi22** [comment]
@simonjayhawkins ??

**Comment by mroeschke** [comment]
Thanks for the pull request, but it appears to have gone stale. If interested in continuing, please merge in the main branch, address any review comments and/or failing tests, and we can reopen.

**Pull request closed without merging**
